### PR TITLE
on_damage EOCs

### PR DESCRIPTION
**Pull request by bombasticSlacks**
#### Summary
None

#### Purpose of change
Request from discord, adds an on damage EOCs for damage types

#### Describe the solution
Added it post absorb where monsters and characters have been handled.
Added documentation.
Added tests.

#### Testing
Unit tests

### CODE REVIEW

File: tests/melee_test.cpp [hunk 1]
```diff
@@ -347,11 +347,41 @@ static void check_damage_from_test_fire( const std::string &mon_id, int expected
...
+        dude.melee_attack( mon, false );
+        if( mon.get_hp() < mon.get_hp_max() ) {
+            total_dmg += mon.get_hp_max() - mon.get_hp();
+            if( mon.get_value( "npctalk_var_general_dmg_type_test_test_fire" ) == "target" ) {
+                REQUIRE( dude.get_value( "npctalk_var_general_dmg_type_test_test_fire" ) == "source" );
+            }
+
+            eoc_total_dmg += std::ceil( std::stod(
+                                            dude.get_value( "npctalk_var_test_damage_taken" ) ) );
+            REQUIRE( dude.get_value( "npctalk_var_test_bp" ) != "" );
```
Comment (bombasticSlacks) [inline tests/melee_test.cpp:378]:
```suggestion
            REQUIRE( !dude.get_value( "npctalk_var_test_bp" ).empty() );
```

File: tests/melee_test.cpp [hunk 2]
```diff
@@ -347,11 +347,41 @@ static void check_damage_from_test_fire( const std::string &mon_id, int expected
...
+        REQUIRE( mon.get_value( "npctalk_var_general_dmg_type_test_test_fire" ).empty() );
+        dude.set_wielded_item( item( "test_fire_sword" ) );
+        dude.melee_attack( mon, false );
+        if( mon.get_hp() < mon.get_hp_max() ) {
+            total_dmg += mon.get_hp_max() - mon.get_hp();
+            if( mon.get_value( "npctalk_var_general_dmg_type_test_test_fire" ) == "target" ) {
+                REQUIRE( dude.get_value( "npctalk_var_general_dmg_type_test_test_fire" ) == "source" );
+            }
+
+            eoc_total_dmg += std::ceil( std::stod(
```
Comment (bombasticSlacks) [inline tests/melee_test.cpp:376]:
```suggestion
            eoc_total_dmg += std::round( std::stod(
```

**Comment by Maleclypse** [comment]
```
> -------------------------------------------------------------------------------
(~[slow] ~[.],starting_items)=> Damage_type_EOCs
(~[slow] ~[.],starting_items)=>   Attacking a monster
(~[slow] ~[.],starting_items)=> -------------------------------------------------------------------------------
(~[slow] ~[.],starting_items)=> ../tests/melee_test.cpp:463
(~[slow] ~[.],starting_items)=> ...............................................................................
(~[slow] ~[.],starting_items)=> 
(~[slow] ~[.],starting_items)=> ../tests/melee_test.cpp:382: FAILED:
(~[slow] ~[.],starting_items)=>   CHECK( eoc_total_dmg == total_dmg )
(~[slow] ~[.],starting_items)=> with expansion:
Error: (~[slow] ~[.],starting_items)=>   8 == 17
(~[slow] ~[.],starting_items)=> 
```
Looks like your new test is failing on it's second run.

**Comment by bombasticSlacks** [comment]
> Looks like your new test is failing on it's second run.

Just haven't had the time to fix it

**Pull request merged**